### PR TITLE
LieAlgebras: further improvements and new features

### PR DESCRIPTION
**Pull request by lgoettgens**
This includes:

- type stability fixes
- general cleanup
- replacing `base_ring` by `coefficient_ring`
- adding `universal_enveloping_algebra` as a special constructor for `PBWAlgebra`
- pretty printing
- new type: ideals of Lie algebras
- new type: Lie subalgebras
- some important ideals and subalgebras (center, derived algebra,...)

### CODE REVIEW

**Comment by lgoettgens** [comment]
The errors on nightly are not due to changes here.

**Comment by thofma** [comment]
@lgoettgens Do you know who would be a suitable reviewer?

**Comment by lgoettgens** [comment]
I didn't see I can request reviews now by myself. I chose @fingolfin and @ThomasBreuer as they are, by my knowledge, the two people from the OSCAR team having a background and maybe an opinion on Lie algebras.

File: experimental/LieAlgebras/src/LieAlgebraModule.jl [hunk 1]
```diff
@@ -836,26 +867,27 @@ function symmetric_power(V::LieAlgebraModule{C}, k::Int) where {C<:RingElement}
     basis_change_T2S * transformation_matrix(T, i) * basis_change_S2T
   end
 
-  if k == 1
-    s = symbols(V)
+  s = if k == 1
+    symbols(V)
+  elseif is_standard_module(V)
+    s = [
```
Comment (ThomasBreuer) [inline experimental/LieAlgebras/src/LieAlgebraModule.jl]: why `s =`?

File: experimental/LieAlgebras/src/LieAlgebraModule.jl [hunk 2]
```diff
@@ -836,26 +867,27 @@ function symmetric_power(V::LieAlgebraModule{C}, k::Int) where {C<:RingElement}
...
+        ),
+      ) for inds in ind_map
+    ]
   else
-    if is_standard_module(V)
-      parentheses = identity
-    else
-      parentheses = x -> "($x)"
-    end
     s = [
```
Comment (ThomasBreuer) [inline experimental/LieAlgebras/src/LieAlgebraModule.jl]: why `s =`?

File: experimental/LieAlgebras/src/LieSubalgebra.jl [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+  @req base_lie_algebra(S1) === base_lie_algebra(S2) "Incompatible Lie algebras."
+  return all(in(S2), gens(S1))
+end
+
+function Base.:(==)(
+  S1::LieSubalgebra{C,LieT}, S2::LieSubalgebra{C,LieT}
+) where {C<:RingElement,LieT<:LieAlgebraElem{C}}
+  base_lie_algebra(S1) === base_lie_algebra(S2) || return false
+  gens(S1) == gens(S2) && return true
+  return basis_matrix(S1) == basis_matrix(S2)
```
Comment (ThomasBreuer) [inline experimental/LieAlgebras/src/LieSubalgebra.jl]: Really? If the subalgebras have been constructed with `is_basis == true` then just the given bases are stored.
Are they necessarily equal if they span the same subalgebra?
(If yes then perhaps the term `is_basis` should perhaps be changed to `is_canonical_basis`.)

File: experimental/LieAlgebras/src/LieSubalgebra.jl [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+#
+###############################################################################
+
+@doc raw"""
+    lie_algebra(S::LieSubalgebra) -> LieAlgebra
+
+Return `S` as a Lie algebra.
+"""
+function lie_algebra(S::LieSubalgebra)
+  return lie_algebra(basis(S)) #, embedding_hom   # TODO
```
Comment (ThomasBreuer) [inline experimental/LieAlgebras/src/LieSubalgebra.jl:260]: There are other places, too, where an independent new algebraic structure gets created from a given substructure. Either the embedding morphism should be returned by the function that constructs the new object, or a function `is_subalgebra` (or so) should be provided that computes the embedding on demand (similar to the situation with `is_subfield` or `is_subgroup`).

File: experimental/LieAlgebras/src/LieAlgebraIdeal.jl [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+#
+###############################################################################
+
+@doc raw"""
+    normalizer(L::LieAlgebra, I::LieAlgebraIdeal) -> LieSubalgebra
+
+Return the normalizer of `I` in `L`, i.e. $\{x \in L \mid [x, I] \subseteq I\} = L$.
+"""
+function normalizer(L::LieAlgebra, I::LieAlgebraIdeal)
+  return sub(L)
```
Comment (ThomasBreuer) [inline experimental/LieAlgebras/src/LieAlgebraIdeal.jl:207]: The code says that the normalizer of an ideal is the full Lie algebra, the documentation states the definition of a normalizer; this looks strange.

It should be checked that `I` is really an ideal in `L`.

File: experimental/LieAlgebras/src/LieAlgebraIdeal.jl [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+  @req base_lie_algebra(I1) === base_lie_algebra(I2) "Incompatible Lie algebras."
+  return all(in(I2), gens(I1))
+end
+
+function Base.:(==)(
+  I1::LieAlgebraIdeal{C,LieT}, I2::LieAlgebraIdeal{C,LieT}
+) where {C<:RingElement,LieT<:LieAlgebraElem{C}}
+  base_lie_algebra(I1) === base_lie_algebra(I2) || return false
+  gens(I1) == gens(I2) && return true
+  return basis_matrix(I1) == basis_matrix(I2)
```
Comment (ThomasBreuer) [inline experimental/LieAlgebras/src/LieAlgebraIdeal.jl]: (See the remark about equality of subalgebras.)

File: experimental/LieAlgebras/src/LieAlgebra.jl [hunk 7]
```diff
@@ -215,39 +211,39 @@ end
...
 function Base.:*(x::LieAlgebraElem{C}, c::C) where {C<:RingElem}
-  base_ring(x) != parent(c) && error("Incompatible rings.")
-  return parent(x)(Generic._matrix(x) * c)
+  coefficient_ring(x) != parent(c) && error("Incompatible rings.")
+  return parent(x)(_matrix(x) * c)
 end
 
 function Base.:*(
   x::LieAlgebraElem{C}, c::U
 ) where {C<:RingElement,U<:Union{Rational,Integer}}
```
Comment (ThomasBreuer) [inline experimental/LieAlgebras/src/LieAlgebra.jl]: Is there a reason not to include `IntegerUnion`?

File: experimental/LieAlgebras/src/LieAlgebra.jl [hunk 8]
```diff
@@ -215,39 +211,39 @@ end
...
 
 function Base.:*(c::C, x::LieAlgebraElem{C}) where {C<:RingElem}
-  base_ring(x) != parent(c) && error("Incompatible rings.")
-  return parent(x)(c * Generic._matrix(x))
+  coefficient_ring(x) != parent(c) && error("Incompatible rings.")
+  return parent(x)(c * _matrix(x))
 end
 
 function Base.:*(
   c::U, x::LieAlgebraElem{C}
```
Comment (ThomasBreuer) [inline experimental/LieAlgebras/src/LieAlgebra.jl]: Omit `{C}`?

File: experimental/LieAlgebras/src/LieAlgebraModule.jl [hunk 9]
```diff
@@ -488,7 +534,7 @@ end
 
 Check whether `V` is a direct sum of modules.
```
Comment (ThomasBreuer) [inline experimental/LieAlgebras/src/LieAlgebraModule.jl]: Perhaps mention that this means whether `V` has been constructed as a direct sum, not whether `V` can be decomposed into a direct sum of submodules (same for `is_tensor_product`, `is_exterior_power`, ...)?

**Review by ThomasBreuer — APPROVED** [review]
This looks overall very nice.
I have left a few minor comments.

File: experimental/LieAlgebras/src/LieSubalgebra.jl (hunk 3 above)
Comment (lgoettgens) [reply experimental/LieAlgebras/src/LieSubalgebra.jl]: This implementation of `==` is just testing equality as julia objects, not of mathematical objects. A change is added in 2843a5b686aa814a0522fac15b5f0149614e5449.

File: experimental/LieAlgebras/src/LieAlgebraIdeal.jl (hunk 6 above)
Comment (lgoettgens) [reply experimental/LieAlgebras/src/LieAlgebraIdeal.jl]: see above. same change applied as for subalgebras

File: experimental/LieAlgebras/src/LieAlgebraModule.jl (hunk 9 above)
Comment (lgoettgens) [reply experimental/LieAlgebras/src/LieAlgebraModule.jl]: Changed in 1e40cb835cd2452e348c553d951d4ca2abd8e7e8

File: experimental/LieAlgebras/src/LieAlgebraIdeal.jl (hunk 5 above)
Comment (lgoettgens) [reply experimental/LieAlgebras/src/LieAlgebraIdeal.jl:207]: For subalgebras, the corresponding function actually calculates something. As the normalizer of an ideal always is the full Lie algebra, there is nothing to check here. I adapted the docstring to reflect that in 1e40cb835cd2452e348c553d951d4ca2abd8e7e8

File: experimental/LieAlgebras/src/LieSubalgebra.jl (hunk 4 above)
Comment (lgoettgens) [reply experimental/LieAlgebras/src/LieSubalgebra.jl:260]: As reflected by the comment in the code, this is planned for a future PR. As one would first introduce a new type of morphisms, this seemed to be too involved for this PR, but will be added in the future.

File: experimental/LieAlgebras/src/LieAlgebraIdeal.jl (hunk 5 above)
Comment (ThomasBreuer) [reply experimental/LieAlgebras/src/LieAlgebraIdeal.jl:207]: Thanks.
It is still possible that someone enters an `I` that is an ideal in some other Lie algebra, not in `L`.

Comment (lgoettgens) [reply experimental/LieAlgebras/src/LieAlgebraIdeal.jl:207]: Oh yeah, that's true. I will add a check for that after lunch

Comment (lgoettgens) [reply experimental/LieAlgebras/src/LieAlgebraIdeal.jl:207]: Added in e7f653429c23a0070ea08f84c19924617d85134e

**Comment by lgoettgens** [comment]
All comments by @ThomasBreuer should be addressed now.

**Comment by ThomasBreuer** [comment]
@lgoettgens Thanks.
There was one test failure which looks not related to this pull request. (I got the same failure in #2410.)

**Pull request merged**